Add support for BigQuery resumable uploads via a write channel

mziccard opened the pull request:
This PR adds support to BigQuery for inserts that use [resumable upload](https://cloud.google.com/bigquery/loading-data-post-request#resumable) using a write channel interface (different from #534).

Main changes are:
- Move `BlobWriteChannel` and `BlobReadChannel` to core module (the latter for consistency) 
- Rename `BlobWriteChannel` and `BlobReadChannel` to `WriteChannel` and `ReadChannel`
- Add abstract class `BaseWriteChannel `implementing entity-agnostic channel functionality
- Add `BlobWriteChannel` and `BlobReadChannel` implementations to `gcloud-java-storage`
- Add `LoadConfiguration` and modify `LoadJobInfo` to take configuration as a parameter
- Add `BigQuery.writer` method to return a write channel given a `LoadConfiguration`
- Add `BigQueryRpc.open` and `.write` methods to implement write channel
- Add `TableDataWriteChannel` class to support bigquery resumable streaming inserts
- Add unit and integration tests
- Update bigquery example with load-data action

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+import java.util.Objects;
+
+/**
+ * Google BigQuery Configuration for a Load operation. A Load configuration can be used to build a
```
Nit: should "Load" --> "load?  It looks like "load" isn't capitalized in the BigQuery docs (e.g. in step 2 [here](https://cloud.google.com/bigquery/loading-data-into-bigquery#loaddatagcs))

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+
+    /**
+     * Sets whether BigQuery should allow extra values that are not represented in the table schema.
+     * If {@code true}, the extra values are ignored. If {@code true}, records with extra columns
```
Should the second {@code true} be "false" instead?

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+  /**
+   * Returns the action that should occur if the destination table already exists.
+   *
+   * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs#configuration.load">
```
Nit: I think it would slightly clearer to link to the write disposition entry directly (#configuration.load.writeDisposition).  Same for `createDisposition()` and both methods' equivalents in the builder.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+  /**
+   * Returns the format of the data files.
+   */
+  public String format() {
```
Should this be `formatOptions`?  Also I like the javadoc in the builder for this functionality, can we copy it over for this method?

mziccard replied:
This method returns the format as a string, not the options. To get options you can use `csvOptions()` with returns a `CsvOptions` object (which is the only format options supported for the moment).

ajkannan commented on gcloud-java-core/src/main/java/com/google/gcloud/WriteChannel.java:
```diff
@@ -0,0 +1,48 @@
+  /**
+   * Captures the write channel state so that it can be saved and restored afterwards. The original
+   * {@code WriteChannel} and the restored one should not both be used. Closing one channel
+   * causes the other channel to close, subsequent writes will fail.
```
Nit: replace the comma with a period or semicolon

ajkannan commented on gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java:
```diff
@@ -0,0 +1,218 @@
+     * @param <EntityT> the entity this channel writes data to. Possibly with additional
+     *     configuration
+     */
+    public abstract static class Builder<
```
I don't think this is tested (correct me if I'm wrong).

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigQueryRpc.java:
```diff
@@ -185,4 +186,26 @@ GetQueryResultsResponse getQueryResults(String jobId, Map<Option, ?> options)
+   * @param uploadId the resumable upload session URI
+   * @param toWrite a byte array of data to upload
+   * @param toWriteOffset offset in the {@code toWrite} param to start writing from
+   * @param destOffset offset in the destition where to upload data to
```
s/destition/destination

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigQueryRpc.java:
```diff
@@ -185,4 +186,26 @@ GetQueryResultsResponse getQueryResults(String jobId, Map<Option, ?> options)
+   * @param toWrite a byte array of data to upload
+   * @param toWriteOffset offset in the {@code toWrite} param to start writing from
+   * @param destOffset offset in the destition where to upload data to
+   * @param length the number of butes to upload
```
s/butes/bytes

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigQueryRpc.java:
```diff
@@ -185,4 +186,26 @@ GetQueryResultsResponse getQueryResults(String jobId, Map<Option, ?> options)
+   * @param toWriteOffset offset in the {@code toWrite} param to start writing from
+   * @param destOffset offset in the destition where to upload data to
+   * @param length the number of butes to upload
+   * @param last if {@code true} indicates that the last chunk if being uploaded
```
The second "if" (near the end of the line) should be an "is".  This also might be clearer if worded as "true indicates that the ..." (removing the "if")

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -64,6 +76,9 @@
   private static final Set<Integer> RETRYABLE_CODES = ImmutableSet.of(500, 502, 503, 504);
+  private static final String BASE_RESUMABLE_URI =
+      "https://www.googleapis.com/upload/bigquery/v2/projects/";
+  private static final int HTTP_RESUME_INCOMPLETE = 308;
```
Do you mind adding a link for this error code as a comment?  It doesn't see to be on the same table as the others, and it'd be good to have it documented.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -417,4 +432,69 @@ public QueryResponse query(QueryRequest request) throws BigQueryException {
+          new ByteArrayContent(null, toWrite, toWriteOffset, length));
+      long limit = destOffset + length;
+      StringBuilder range = new StringBuilder("bytes ");
+      range.append(destOffset).append('-').append(limit - 1).append('/');
```
question: why is the last '/' necessary?

mziccard commented:
@ajkannan Thanks for the feedback, I fixed the issues in place

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -417,4 +433,69 @@ public QueryResponse query(QueryRequest request) throws BigQueryException {
+        message = ex.getStatusMessage();
+      }
+      if (!last && code != HTTP_RESUME_INCOMPLETE
+          || last && !(code == HTTP_OK || code == HTTP_CREATED)) {
```
It seems like there's a [very small] chance for throwing an exception when the upload actually worked (under step 2 in the [example](https://cloud.google.com/bigquery/loading-data-post-request#resume-upload)).  Do you think this is worth documenting?

mziccard replied:
I am missing how this can happen. The request for status must have content range of the form `Content-Range: bytes */<limit>`. If I am not mistaken this can never occur as we always set a range.

ajkannan replied:
You're right (I confused myself earlier); ignore my original comment.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -417,4 +433,69 @@ public QueryResponse query(QueryRequest request) throws BigQueryException {
+  }
+
+  @Override
+  public void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset, int length,
```
Is it possible to overwrite bytes with a request?  I would guess yes, but I figured I'd ask anyways just because I didn't see that in any example on the resumable upload docs.

mziccard replied:
Yes you should be able to do that. The upload mechanism is the exact same used in storage.

ajkannan commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/LoadConfigurationTest.java:
```diff
@@ -0,0 +1,123 @@
+  @Test
+  public void testToBuilderIncomplete() {
+    LoadConfiguration configuration = LoadConfiguration.of(TABLE_ID);
+    compareLoadConfiguration( configuration, configuration.toBuilder().build());
```
Nit: extra space after '('

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/TableDataWriteChannel.java:
```diff
@@ -0,0 +1,109 @@
+      }
+    }
+
+    static Builder builder(BigQueryOptions options, LoadConfiguration blobInfo, String uploadId) {
```
s/blobInfo/configuration

ajkannan commented:
@mziccard I took one more pass, and LGTM except for one small variable name change.  Feel free to merge after fixing that.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java:
```diff
@@ -0,0 +1,218 @@
+    ServiceOptionsT extends ServiceOptions<?, ?, ServiceOptionsT>,
+    EntityT extends Serializable> implements WriteChannel {
+
+  private static final int MIN_CHUNK_SIZE = 256 * 1024;
```
At some point this may need to be overridden by subclasses. current value is specific to GCS.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+
+    /**
+     * Sets the schema for the destination table. The schema can be omitted if the destination table
+     * already exists, or if you're loading data from Google Cloud Datastore.
```
from Google Cloud Datastore backup (DATASTORE_BACKUP format option)?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/LoadConfiguration.java:
```diff
@@ -0,0 +1,381 @@
+     * all properties. If any named property isn't found in the Cloud Datastore backup, an invalid
+     * error is returned in the job result.
+     */
+    public Builder projectionFields(List<String> projectionFields) {
```
shouldn't it be part of the DATASTORE_BACKUP FormatOptions? I mean via a subclasses, similar to CSV?

mziccard replied:
You are right, I opened a new issue (#552) as this PR is already big enough :)

aozarov commented:
Looks good! I would like to give it another pass tomorrow. Until then few questions/comments. Also,
I think there are still unaddressed comments made by @ajkannan.

aozarov commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java:
```diff
@@ -41,10 +41,12 @@
 import com.google.gcloud.spi.BigQueryRpc;
 import com.google.gcloud.spi.BigQueryRpc.Tuple;
 import com.google.gcloud.spi.BigQueryRpcFactory;
+import com.google.gcloud.storage.BlobInfo;
```
Not needed.

aozarov commented on gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/ITBigQueryTest.java:
```diff
@@ -16,6 +16,7 @@
 
 package com.google.gcloud.bigquery;
 
+import static com.google.gcloud.bigquery.BigQuery.*;
```
replace with specific.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java:
```diff
@@ -0,0 +1,226 @@
+  private static final int MIN_CHUNK_SIZE = 256 * 1024;
+  private static final int DEFAULT_CHUNK_SIZE = 8 * MIN_CHUNK_SIZE;
+
+  protected final ServiceOptionsT options;
```
To prevent bad practice or mishandling in sub-classes I suggest to have all members private with protected accessors.
This is not an issue for the final members but I would favor consistency.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java:
```diff
@@ -0,0 +1,226 @@
+    return DEFAULT_CHUNK_SIZE;
+  }
+
+  protected abstract void write(int length, boolean last);
```
javadoc would be nice. Also maybe name should be more specific (e.g. `writeBuffer`, `flushBuffer`, ...).

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/BaseWriteChannel.java:
```diff
@@ -0,0 +1,226 @@
+    this.uploadId = uploadId;
+  }
+
+  protected void flush() {
```
I would make anything that is not need to be overriden by the subclasses either package scope, private or final (if must be public). Especially methods that would easily get it wrong and require deep understanding on the base class internals.

aozarov commented on gcloud-java-core/src/main/java/com/google/gcloud/ReadChannel.java:
```diff
@@ -0,0 +1,57 @@
+   *     it afterwards.
+   */
+  @Override
+  RestorableState<ReadChannel> capture();
```
I think this could be implemented in `BaseWriteChannel` using an abstract protected method that would return a newly created `BaseState.Builder` instance.

aozarov commented:
Ok, finished with my pass. Looks good. Few more suggestion and some cleanup.
Also, do you understand why the coverage was dropped in this PR by almost 1 percent?

mziccard commented:
@aozarov I fixed your comments and rebased.

> do you understand why the coverage was dropped in this PR by almost 1 percent?

I have been trying to figure this out but have not so many ideas. A decrease might be caused by the fact that this adds code to the Rpc layer which is only tested in ITs that are (I think) not considered by *coveralls* report. However this ~1% seems a bit high to me.
